src/Development/IDE/Core/Shake.hs

```diff
@@ -268,30 +302,55 @@ instance NFData v => NFData (Value v)
 currentValue :: Value v -> Maybe v
 currentValue (Succeeded _ v) = Just v
 currentValue (Stale _ _) = Nothing
-currentValue Failed = Nothing
+currentValue Failed{} = Nothing
 
 -- | Return the most recent, potentially stale, value and a PositionMapping
 -- for the version of that value.
-lastValueIO :: ShakeExtras -> NormalizedFilePath -> Value v -> IO (Maybe (v, PositionMapping))
-lastValueIO ShakeExtras{positionMapping} file v = do
-    allMappings <- liftIO $ readVar positionMapping
-    pure $ case v of
-        Succeeded ver v -> Just (v, mappingForVersion allMappings file ver)
-        Stale ver v -> Just (v, mappingForVersion allMappings file ver)
-        Failed -> Nothing
+lastValueIO :: IdeRule k v => ShakeExtras -> k -> NormalizedFilePath -> IO (Maybe (v, PositionMapping))
+lastValueIO s@ShakeExtras{positionMapping,persistentKeys,state} k file = do
+    let readPersistent = do
+          pmap <- readVar persistentKeys
+          mv <- runMaybeT $ do
+            liftIO $ Logger.logDebug (logger s) $ T.pack $ "LOOKUP UP PERSISTENT FOR" ++ show k
+            f <- MaybeT $ pure $ HMap.lookup (Key k) pmap
+            dv <- MaybeT $ runIdeAction "lastValueIO" s $ f file
+            MaybeT $ pure $ fromDynamic dv
+          modifyVar state $ \hm -> case mv of
+            Nothing -> pure (HMap.insertWith upd (file,Key k) (Failed True) hm,Nothing)
+            Just v -> pure (HMap.insertWith upd (file,Key k) (Stale Nothing (toDyn v)) hm, Just (v,zeroMapping))
```

Comment:
This returns `zeroMapping` because we cannot do better, right?

Reply:
Yes, with a bit more work we can compute the position mapping from the diff.

Reply:
Add a TODO then!